contracts/citycoin-core.clar

```diff
@@ -1,201 +1,539 @@
+(define-data-var activeContract principal CONTRACT_OWNER)
+
+;; defines if the initial logic contract is provided for initialization
+(define-data-var initialized bool false)
 
-;; used as Proposal ID
-(define-data-var proposalNonce uint u0)
+;; updates the activeContract variable with initial logic contract
+;; can only be run once after deployment
+(define-public (setup-active-contract (firstActiveContract principal))
+  (begin
+    (asserts! (is-authorized-owner) (err ERR_UNAUTHORIZED))
+    (asserts! (not (var-get initialized)) (err ERR_UNAUTHORIZED))
+    (var-set activeContract firstActiveContract)
+    (var-set initialized true)
+    (ok true)
+  )
+)
 
-(define-map Contracts
-  principal  ;; MiningContract address
-  { proposalId: uint, name: (string-ascii 100) }
+;; returns active contract
+(define-read-only (get-active-contract)
+  (var-get activeContract)
 )
 
-(define-map Proposals
-  uint ;; Proposal id
+;; store logic contract information
+(define-map CityCoinContracts
+  principal
   {
-    address: principal,
-    startBH: uint,
-    endBH: uint,
-    voters: uint,
-    votes: uint,
-    isOpen: bool
+    state: uint,
+    startHeight: uint,
+    endHeight: uint,
+    active: bool
   }
 )
 
-;; check
-(define-map ProposalVoters
-  { proposalId: uint, voter: principal }
-  bool
+;; returns logic contract information
+(define-read-only (get-contract (address principal))
+  (map-get? CityCoinContracts address)
 )
 
-(define-read-only (get-proposals-count)
-  (var-get proposalNonce)
+;; set initial value for logic contract
+(map-set CityCoinContracts
+  .citycoin-logic-v1
+  {
+    state: CONTRACT_DEFINED,
+    startHeight: u0,
+    endHeight: u0,
+    active: false 
+  })
+
+;; TODO: function to update active contract
+;;   called by register miner to activate first contract
+;;   called by city wallet to update to new contract (+ delay)
+;; (try! (contract-call? .citycoin-logic-v1 startup (some activationBlockVal)))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; REGISTRATION
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(define-data-var activationBlock uint u340282366920938463463374607431768211455)
+(define-data-var activationDelay uint u150)
+(define-data-var activationReached bool false)
+(define-data-var activationThreshold uint u20)
+(define-data-var usersNonce uint u0)
+
+;; returns Stacks block height registration was activated at plus activationDelay
+(define-read-only (get-activation-block)
+  (let
+    (
+      (activated (var-get activationReached))
+    )
+    (asserts! activated (err ERR_CONTRACT_NOT_ACTIVATED))
+    (ok (var-get activationBlock))
+  )
 )
 
-(define-read-only (has-voted-on-proposal (proposalId uint) (who principal))
-  (is-some (map-get? ProposalVoters
-    { proposalId: proposalId, voter: who }
-  ))
+;; returns activation delay
+(define-read-only (get-activation-delay)
+  (var-get activationDelay)
```

Comment:
It depends on the purpose of each read-only function. In some cases it felt like it made more sense to return the variable directly, especially for something that is true/false or easily defaulted to 0 values.

In other cases like `get-coinbase-thresholds` it made sense to check if the contract was activated first and throw an error if not.

There may be room to re-evaluate and add some more consistency here. Curious what you would recommend?

Reply:
Yeah - not sure either. I was just curious if you had any best practices.

For `read-only` functions I prefer just returning the value and not a response (since then you have to unwrap etc) but I'm not sure if that is possible. We also have a mix.

Reply:
`read-only` functions can return any valid data type. And `response` data type is one of them. So there is nothing wrong in returning sometimes pure value, and sometimes a response type (`ok` or `err`)

But maybe names of our functions should be more explicit?

Let's say we have a function `can-mine`. Based on its name it should return either `true` or `false`. But there can be multiple reasons why you might not be allowed to mine and it would be good to tell user why it is so.
So instead of:
```clarity
(define-read-only (can-mine (amount uint))
  (begin
    (asserts! (> block-height u200) (err ERR_ERROR_1))
    (asserts! (> amount u0) (err ERR_ERROR_2))
    (asserts! (>= (stx-get-balance tx-sender) amount) (err ERR_ERROR_3))
    (ok true)
  )
)
```
maybe we should define two functions:
```
(define-read-only (can-mine-or-err (amount uint))
  (begin
    (asserts! (> block-height u200) (err ERR_ERROR_1))
    (asserts! (> amount u0) (err ERR_ERROR_2))
    (asserts! (>= (stx-get-balance tx-sender) amount) (err ERR_ERROR_3))
    (ok true)
  )
)

(define-read-only (can-mine (amount uint))
  (is-ok (can-mine-or-err amount))
)
```
so when you only need boolean value, you can call `can-mine`, but when you are interested also in reason why user might not be allowed to mine, then you can use `can-mine-or-error`.